src/monitor/node_active_protocol.c

```diff
@@ -1281,6 +1282,157 @@ perform_failover(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * promote promotes a given target node in a group.
+ */
+Datum
+perform_promotion(PG_FUNCTION_ARGS)
+{
+	text *formationIdText = PG_GETARG_TEXT_P(0);
+	char *formationId = text_to_cstring(formationIdText);
+
+	text *nodeNameText = PG_GETARG_TEXT_P(1);
+	char *nodeName = text_to_cstring(nodeNameText);
+
+	List *groupNodesList = NULL;
+	int totalNodesCount = 0;
+
+	AutoFailoverNode *currentNode = NULL;
+
+	checkPgAutoFailoverVersion();
+
+	currentNode = GetAutoFailoverNodeByName(formationId, nodeName);
+
+	if (currentNode == NULL)
+	{
+		ereport(ERROR,
+				(errmsg("node \"%s\" is not registered in formation \"%s\"",
+						nodeName, formationId)));
+	}
+
+	LockFormation(formationId, ShareLock);
+	LockNodeGroup(formationId, currentNode->groupId, ExclusiveLock);
+
+	/*
+	 * If the current node is the primary, that's done.
+	 */
+	if (IsCurrentState(currentNode, REPLICATION_STATE_SINGLE) ||
+		IsCurrentState(currentNode, REPLICATION_STATE_PRIMARY))
+	{
+		/* return false: no promotion is happening */
+		ereport(NOTICE,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg("cannot perform promotion: node %s in formation %s "
+						"is already a primary.",
+						nodeName, formationId)));
+		PG_RETURN_BOOL(false);
+	}
+
+	/*
+	 * If the node is not a primary, it needs to be in the SECONDARY state.
+	 * When we call perform_failover() to implement the actual failover
+	 * orchestration, this condition is going to be checked again, but in a
+	 * different way.
+	 *
+	 * For instance, the target could be in MAINTENANCE and perform_failover
+	 * would still be able to implement a failover given another secondary node
+	 * being around.
+	 */
+	if (!IsCurrentState(currentNode, REPLICATION_STATE_SECONDARY))
+	{
+		/* return false: no promotion is happening */
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+				 errmsg(
+					 "cannot perform promotion: node %s in formation %s "
+					 "has reported state \"%s\" and is assigned state \"%s\", "
+					 "promotion can only be performed when "
+					 "in state \"secondary\".",
+					 nodeName, formationId,
+					 ReplicationStateGetName(currentNode->reportedState),
+					 ReplicationStateGetName(currentNode->goalState))));
+	}
+
+	/*
+	 * If we have only two nodes in the group, then perform a failover.
+	 */
+	groupNodesList =
+		AutoFailoverNodeGroup(currentNode->formationId, currentNode->groupId);
+
+	totalNodesCount = list_length(groupNodesList);
+
+	if (totalNodesCount <= 2)
+	{
+		DirectFunctionCall2(perform_failover,
+							CStringGetTextDatum(formationId),
+							Int32GetDatum(currentNode->groupId));
+
+		/* if we reach this point, then a failover is in progress */
+		PG_RETURN_BOOL(true);
+	}
+	else
+	{
+		char message[BUFSIZE] = { 0 };
+
+		/*
+		 * In the general case, we perform a little trick:
```

Comment:
minor: maybe move this comment after the `if (currentNode->candidatePriority == 0)` check is done